nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-processors/src/main/java/org/apache/nifi/processors/elasticsearch/ScrollElasticsearchHttp.java

```diff
@@ -0,0 +1,415 @@
+                this.getPage(getResponse, scrollurl, context, session, flowFile, logger, startNanos);
+            } else {
+                logger.debug("Querying {}/{} from Elasticsearch: {}", new Object[] { index,
+                        docType, query });
+
+                // read the url property from the context
+                final String urlstr = StringUtils.trimToEmpty(context.getProperty(ES_URL)
+                        .getValue());
+                final URL queryUrl = buildRequestURL(urlstr, query, index, docType, fields, sort,
+                        scrollId, pageSize, scroll);
+                final long startNanos = System.nanoTime();
+
+                final Response getResponse = sendRequestToElasticsearch(okHttpClient, queryUrl,
+                        username, password, "GET", null);
+                this.getPage(getResponse, queryUrl, context, session, flowFile, logger, startNanos);
+            }
+
+        } catch (IOException ioe) {
+            logger.error(
+                    "Failed to read from Elasticsearch due to {}, this may indicate an error in configuration "
+                            + "(hosts, username/password, etc.).",
+                    new Object[] { ioe.getLocalizedMessage() }, ioe);
+            session.remove(flowFile);
+            context.yield();
+
+        } catch (Exception e) {
+            logger.error("Failed to read {} from Elasticsearch due to {}", new Object[] { flowFile,
+                    e.getLocalizedMessage() }, e);
+            session.transfer(flowFile, REL_FAILURE);
+            context.yield();
+        }
+    }
+
+    private void getPage(final Response getResponse, final URL url, final ProcessContext context,
+            final ProcessSession session, FlowFile flowFile, final ComponentLog logger, final long startNanos)
+            throws IOException {
+        final int statusCode = getResponse.code();
+
+        if (isSuccess(statusCode)) {
+            ResponseBody body = getResponse.body();
+            final byte[] bodyBytes = body.bytes();
+            JsonNode responseJson = parseJsonResponse(new ByteArrayInputStream(bodyBytes));
+            String scrollId = responseJson.get("_scroll_id").asText();
+
+            StringBuilder builder = new StringBuilder();
+
+            builder.append("{ \"hits\" : [");
+
+            JsonNode hits = responseJson.get("hits").get("hits");
+            if (hits.size() == 0) {
+                finishQuery(context.getStateManager());
+                session.remove(flowFile);
+                return;
+            }
+
+            for(int i = 0; i < hits.size(); i++) {
+                JsonNode hit = hits.get(i);
+                String retrievedIndex = hit.get("_index").asText();
+                String retrievedType = hit.get("_type").asText();
+
+                JsonNode source = hit.get("_source");
+                flowFile = session.putAttribute(flowFile, "es.index", retrievedIndex);
```

Comment:
In the Scroll processor, it always returns a full page of JSON documents per flow file, so there isn't really an easy way to include the id in an attribute (though I suppose we could so something like es.result.id.1=[id], es.result.id.2=[id2], etc)